File: java/src/Bj1780.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Bj1780 {
    static void solve(){
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        try {
            int n = Integer.parseInt(br.readLine());
            int[][] arr = new int[n][n];
            for (int i = 0; i < n; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int j = 0; j < n; j++) {
                    arr[i][j] = Integer.parseInt(st.nextToken());
                }
            }
            int[] ans = {0, 0, 0};
            cut(arr, ans, 0, 0, n);
            System.out.println(ans[0]);
            System.out.println(ans[1]);
            System.out.println(ans[2]);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void cut(int[][] arr, int[] ans, int x, int y, int n) {
        int index = arr[x][y];
        boolean flag = false;
        for (int i = x; i < x + n; i++) {
            for (int j = y; j < y + n; j++) {
                if (index != arr[i][j]) {
                    int cutting = n / 3;
                    for (int stepX = 0; stepX < 3; stepX++) {
                        for (int stepY = 0; stepY < 3; stepY++) {
                            cut(arr, ans, x + stepX * cutting, y + stepY * cutting, cutting);
                        }
                    }
                    flag = true;
                    break;
                }
            }
            if (flag) {
                break;
            }
        }
        if (!flag) {
            switch (index) {
                case -1:
                    ans[0] += 1;
                    break;
                case 0:
                    ans[1] += 1;
                    break;
                case 1:
                    ans[2] += 1;
                    break;
            }
        }
    }
}

File: kotlin/src/Bj1780.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class Bj1780 {
    companion object{
        fun solve(){
            val br = BufferedReader(InputStreamReader(System.`in`))
            val n = br.readLine().toInt()
            val arr = Array(n) { Array(n) { 0 } }
            for (i in 0 until n) {
                val st = StringTokenizer(br.readLine())
                for (j in 0 until n) {
                    arr[i][j] = st.nextToken().toInt()
                }
            }
            val ans = arrayOf(0, 0, 0)
            cut(arr, ans, 0, 0, n)
            println(ans[0])
            println(ans[1])
            println(ans[2])
        }

        private fun cut(arr: Array<Array<Int>>, ans: Array<Int>, x: Int, y: Int, n: Int) {
            val index = arr[x][y]
            var flag = false
            for (i in x until x + n) {
                for (j in y until y + n) {
                    if (index != arr[i][j]) {
                        val cutting = n / 3
                        for (stepX in 0 until 3) {
                            for (stepY in 0 until 3) {
                                cut(arr, ans, x + stepX * cutting, y + stepY * cutting, cutting)
                            }
                        }
                        flag = true
                        break
                    }
                }
                if (flag) {
                    break
                }
            }
            if (!flag) {
                when (index) {
                    -1 -> ans[0] += 1
                    0 -> ans[1] += 1
                    1 -> ans[2] += 1
                }
            }
        }
    }
}